File: 234-palindrome-linked-list/go/main.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func isPalindrome(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return true
	}

	slow := head
	fast := head.Next

	for fast != nil && fast.Next != nil {
		fast = fast.Next.Next
		slow = slow.Next
	}

	if fast == nil {
		temp := head
		for temp.Next != slow {
			temp = temp.Next
		}
		temp.Next = nil
	}

	head2 := reverseList(slow.Next)
	slow.Next = nil

	for head != nil || head2 != nil {
		if head == nil || head2 == nil || head.Val != head2.Val {
			return false
		}
		head = head.Next
		head2 = head2.Next
	}

	return head == head2
}

func reverseList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	a := head
	b := a.Next
	c := b.Next

	a.Next = nil

	for b != nil {
		b.Next = a
		a = b
		b = c
		if c != nil {
			c = c.Next
		}
	}

	return a
}

File: 234-palindrome-linked-list/cpp/main.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
 public:
  bool isPalindrome(ListNode* head) {
    if (head == NULL || head->next == NULL) {
      return true;
    }
    auto slow = head;
    auto fast = head->next;

    while (fast != NULL && fast->next != NULL) {
      fast = fast->next->next;
      slow = slow->next;
    }

    if (fast == NULL) {
      auto temp = head;
      while (temp->next != slow) {
        temp = temp->next;
      }
      temp->next = NULL;
    }
    auto head2 = reverseList(slow->next);
    slow->next = NULL;

    while (head != NULL || head != NULL) {
      if (head == NULL || head2 == NULL) {
        return false;
      }
      if (head->val != head2->val) {
        return false;
      }
      head = head->next;
      head2 = head2->next;
    }

    return head == NULL && head2 == NULL;
  }

  ListNode* reverseList(ListNode* head) {
    if (head == NULL || head->next == NULL) {
      return head;
    }

    auto a = head;
    auto b = head->next;
    auto c = head->next->next;
    a->next = NULL;
    while (b != NULL) {
      b->next = a;
      a = b;
      b = c;
      if (c != NULL) {
        c = c->next;
      }
    }

    return a;
  }
};